### ISSUE TITLE
Feature Request: Low-overhead API

### ISSUE THREAD

**Issue by natejgardner** [arangodb/arangodb#9737]
REST calls are okay for transactional requests, but due to the overhead associated with them, they're not ideal for high volume or low latency applications. 

For example, a single-read `return document('mycollection/mydocument')` query internally takes only 380 us to execute. But via REST call, the performance the client must wait over 300 ms from the time the query is sent to the time the data is received on localhost, three orders of magnitude decrease in effective performance (measured in Chrome dev tools). This looks like it could be minimized to 6 ms if given proper scheduler priority, but even so, this is 15x longer than the query itself takes to run.

ArangoDB's performance is so excellent that given a proper IPC mechanism it could be a very powerful tool with significant advantages over self-implemented data structures or the use of complex libraries for keeping track of application state for time-constrained applications (accessing / updating state in sub-10ms loops, for example). I realize this isn't the typical use-case for ArangoDB, but just by exposing some fast IPC mechanism for local access such as a shared memory interface (e.g. Boost.Interprocess), pipes, sockets, etc., and/or a lower-overhead networked interface, applications with tight time requirements could take advantage of ArangoDB's optimized data structures. It's possible patterns other than req/res could be made available this way as well. Systems like redis, memcached, aerospike, geode, etc. are very limited in terms of complexity of data in exchange for great performance. But ArangoDB provides similar performance with support for much higher complexity. I think ArangoDB could be competitive in the in-memory cache/db space given low-overhead IPC.

**Comment by graetzer** [comment]
Theoretically you can use arango over Unix sockets, additionally you can use the silent option and velocystream.
Unfortunately there is no way to get arangodb to not answer at all to a request, right now

**Comment by jsteemann** [comment]
@natejgardner: are you saying that you are seeing a 0.38 ms query response time for the actual query when run inside the server, and a response time of 300ms when running tge same query from a client appkication?
The latter seems a bit too high on localhost, so I am wondering what you are actually measuring. Can you elaborate a bit on what exactly you are taking the timings of?
Thanks!

**Comment by natejgardner** [comment]
@jsteemann, I just used Chrome dev tools in the ArangoDB web interface. I wrote the above query, opened the Chrome dev tools network tab, submitted the query, and compared the time as measured by ArangoDB internally to the time as measured by Chrome dev tools. I'm sure a lot of factors affected the performance of the request since there was a huge amount of idle time when other Chrome threads were working. Adding up the time actually spent processing the request, ignoring the idle time, the request took 6ms. But even in this case, if I'm on a 100 Hz loop, 6ms awaiting the interface doesn't leave much time for processing and feels quite wasteful when the DB is capable of doing the work in significantly less than 1 ms.

**Comment by jsteemann** [comment]
@natejgardner : ok, got it.

But to be honest, I think the web interface is actually not very suitable for performing response time measurements, for the following reasons:
- the queries from the web interface are posted to the HTTP endpoint for query execution (`/_api/cursor`), but using a special header "x-arango-async", which will queue them on the server and just instantly return a job id. The query will then be started on the server in the background and the web interface can go on doing other things in the meantime. Most importantly, it doesn't block waiting for the query response.
- the web interface then does an instant follow-up request to query the status for the particular job it just got the id for, using the job API at `/_api/job`. In case there is no result ready yet (very likely), it will go to sleep for a considerable amount of time and reschedule a check request. I think it will go to sleep for 500ms. Eventually it will make a successful request to the jobs API and find that the result is ready, and with that fetch the result. But it may have gone to sleep several times before that. This just overexaggerates response times a great deal.

That means the actual response times for queries may actually be a lot lower than when using the web interface. My assumption here is that normally you will be using client drivers and not the web interface of ArangoDB to run queries.

Here is a more compact comparison of query processing using the web interface and a client driver:

*web interface*
1) web interface sends request to /_api/cursor, with special "x-arango-sync" header
2) server instantly responds with job and queues the query on the server
3) web interface queries job status by sending request to /_api/job using the id
4) server responds with "job not finished yet"
5.x) web interface goes to sleep (500 ms)
5.y) server starts executing the query
6) server finishes execution of the query
7) web interface wakes up and queries job status by sending request to /_api/job using the id
8) server responds with "job finished" and pushes results
9) web interface renders query result

*client driver*
1) client driver sends request to /_api/cursor, and blocks
2) server starts executing the query
3) server finishes execution of the query
4) server responds with "query finished" and pushes results

As you can see, query processing will have a lot less overhead when done via a client driver. It is only in the web interface that it adds this extra delays and async retries in order to keep the web interface responsible and not block on waiting for the server's response.

So I suggest trying the query performance with a client driver and check if it is getting better for you too. My guess is that it will, and that it is only the "random" delays caused by the web interface's implementation that cause the significant slowdown.

**Comment by jsteemann** [comment]
I agree that passing messages via shared memory should result in less overhead and thus lower latency.

On the server side this would require a completely new interface and also a new message format, because right now all the server methods expect incoming data (via TCP or Unix domain sockets) to be structured like HTTP requests, with individual headers and a body containing the payload. But basically it is just a big buffer of `char`s that is parsed and decomposed into its individual components.

So adding this to the server is non-trivial already, and then we of course would need to adapt the existing client tools (arangosh, arangoimport, ...) plus client drivers.

Overall this would be a substantial endeavour, so it needs proper analysis and assessment by our product management. @ifcologne
